### ISSUE TITLE
Enforce timeout on rolling back failed transactions

### ISSUE THREAD

**Issue by istreeter** [snowplow/snowplow-scala-analytics-sdk#125]
In #1161 we added a timeout to rollback statements and commit statements.  This was supposed to address an issue in which `SQLTimeoutException` on a statement would lead to the loader getting stuck while trying to roll back.

However, since that code change we have still seen occasions where the loader got stuck after a SQLTimeoutException.

The logs give us these clues:

- At 08:14:48 the loader starts loading a particular table for a context
- Between 08:14:48 and 09:14:48, the StateMonitoring loop occasionally logs a message loading is ongoing for that table.
- At 09:14:48 there is an exception showing a SQLTimeoutException for that context.  This corresponds to the 60 minute timeout we allow for the JDBC statement.
- At 09:31:11 we see a final message from StateMonitoring that loading is ongoing.  This corresponds to the 20 minutes timeout we allow for the transaction to roll back after failure.
- Between 09:31:11 and 21:48:20 there is no activity from the loader and no log messages from StateMonitoring
- At 21:48:20 we finally get this exception in the logs generated by StateMonitoring:

> ERROR StateMonitoring: Loader is stuck. Ongoing processing of s3://xxxxxxxxx/run=2023-02-01-07-46-14/ at copying into xyz_context_1 table. Spent 81 minutes at this statement

Note the error message was generated 81 minutes after loading started (read it!!), but we only saw the exception after 12 hours of no activity.  This must mean the StateMonitoring process got stuck [around this line](https://github.com/snowplow/snowplow-rdb-loader/blob/5.3.1/modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/StateMonitoring.scala#L52), when cleaning up after racing the StateMonitoring background fiber and the loading fiber.  Under the hood, cats-effect [calls `fiber.cancel` on the "losing" fiber](https://github.com/typelevel/cats-effect/blob/v2.5.4/core/shared/src/main/scala/cats/effect/Concurrent.scala#L288), and in our case that means it got stuck when cancelling the fiber doing the load.

...and that must mean it got stuck running a finalizer somewhere, because Cats Effect guarantees that finalizers are allowed to run to completion.  In our case I assume it is [the finalizer of this doobie resource used for jdbc transactions](https://github.com/tpolecat/doobie/blob/v0.13.4/modules/core/src/main/scala/doobie/util/transactor.scala#L167).  This is the finalizer that runs rollback on the transaction.

To summarize... so far we tried using JDBC statement timeouts to prevent hanging on a rollback.  But this approach clearly has not worked.  My next best suggestion is to put a timeout on the finalizer, by using a cats-effect `.timeout` when cancelling the loading fiber.  This definitely is not elegant, because it means we give up and abandon the un-finalized transaction.  But it's better than having a loader that hangs and does nothing useful for 12 hours.

**Comment by istreeter** [comment]
Whoops, opened this in the wrong repo 🤦‍♂️
